Datastructures: Stacks & Linked Lists

RemieRichards opened the pull request:
## General changes:
- Adds the /__DATASTRUCTURES folder
- Moves PriorityQueue data structure out of the AStar file, and into it's own file in /__DATASTRUCTURES
- Adds Stacks, a data structure, to /__DATASTRUCTURES
- Adds Linked Lists, a data structure, to /__DATASTRUCTURES
- Deletes `stack.dm` and the merges the implementation of NTSL's stacks into this version of stacks (for organisational sake and because mine had more procs and features)
## Stack:

A stack is a First-in-Last-Out data structure, Elements are accessed with the `pop()` and `push()` methods primarily, `pop()` returns the last element in the stack, and removes it from the stack (there is also `peek()` which returns the last element without removing it), `push()` adds the element on to the end of the stack, making it the element returned by the next `pop()`
## Linked List:

Linked List: a double-linked linked list, which stores `linked_nodes`s which contain a reference to the node before them (or null), the node after them (or null), the `linked_list` they belong to and the value they hold. Thanks to @Giacom and @MrStonedOne for pointing out the oddity of my original implementation.

I was bored so I implemented Stacks and Linked Lists (double linked lists) in Byond, woo.
## Performance:

<B>1 Million calls</B> to <B>Stack Pop()</B>, was around <B>1CPU</B>
<B>1 Million calls</B> to <B>Stack Push()</B>, was around <B>4CPU</B>
<B>1 Million calls</B> to <B>Linked List Add()</B>, was around <B>3CPU</B>
<B>1 Million calls</B> to <B>Linked List Remove()</B>, was around <B>4CPU</B>

RemieRichards committed:
Datastructures!

* Moves PriorityQueue data structure out of the AStar file, and into it's own file in /__DATSTRUCTURES
* Adds Stacks, a data structure, to /__DATASTRUCTURES
* Adds Linked Lists, a data structure, to /__DATASTRUCTURES

MrPerson commented:
What possible use do linked lists have that couldn't be replicated with regular byond lists?

RemieRichards commented:
@MrPerson never said they'd be useful, I'm just offering them up since I spent 2 hours debugging them.

MrStonedOne commented:
linked lists gives you an ability to send a list to a proc but starting at a certain point, and it can just keep using *next

Iamgoofball commented on code/__DATASTRUCTURES/stacks.dm:
```diff
@@ -0,0 +1,55 @@
+#define STACK_UNDERFLOW	null
+
+
+/datum/stack
```
somewhere sometime someone is going to make a datum related to material stacks named this, and get very confused why shit breaks

RemieRichards replied:
Tbh, Datastructures > Material stacks (that logically should just get called ```material_stack``` anyways :P)

PJB3005 replied:
There already ARE stack datums, actually.

[honkhonkhonkhonk](https://github.com/tgstation/-tg-station/blob/master/code/modules/scripting/stack.dm)

duncathan commented on code/__DATASTRUCTURES/priority_queue.dm:
```diff
@@ -0,0 +1,65 @@
+
+//removes and returns the first element in the queue
+/PriorityQueue/proc/Dequeue()
+	if(!L.len)
```
there is a proc that does this 14 lines above this
is this not what it would be used for???

vista- replied:
Making an extra function call is a performance trade-off with little to no gain in this case.

Miauw62 commented:
I have a suspicion that these are going to be horribly, horribly slow (because of BYOND)

Chnkr commented:
@Miauw62 i had the same suspicion, especially the stack, and the linked lists, byond is horrible at lists, and especially adding and removing things from a list.

RemieRichards commented:
@Miauw62 
@Chnkr 

1 million calls to Stack ```Pop()```, was around 3CPU
1 million calls to Stack ```Push()```, was around 7CPU
1 Million calls to Linked List ```add()```, was around 20CPU
1 Million calls to Linked List ```remove()```, was around 10CPU

The only way you're going to lag things is if you make stupid code, which is no different from how things are now.

@Chnkr as to your specific point on removing things, I was able to get ALL stack related removals to use ```Cut(i,i+1)``` which is the fastest method (fastest random access method) for removing an element from a list.  Sadly LL DO use the normal list ```Remove()``` instead of ```Cut()```, simply because I don't have the index immediately available, and I suspect a ```Find()``` call to get that Index would just defeat the whole purpose.

tl;dr everything is fine.

MrStonedOne commented:
Its kinda odd to do a linked list that is also stored as an array

Ideally the contents isn't needed other than to make it work with a for, but you lose performance gains in linked lists by doing that. and at that point its just an array with extra overhead.

Giacom commented on code/__DATASTRUCTURES/linked_lists.dm:
```diff
@@ -0,0 +1,173 @@
+//Ok so it's technically a double linked list, bite me.
+
+/datum/linked_list
+	var/list/contents = list()
```
This isn't a real linked list, it is more like a [ring buffer](http://gameprogrammingpatterns.com/event-queue.html#a-ring-buffer) with the front and back pointers.

A real linked list would not have an underlying array (list), it would have nodes which point to nodes; while keeping track of the front and (maybe) back node.

RemieRichards replied:
I can do that, gives me an excuse to redo the Front/Back naming scheme to.

Miauw62 replied:
to the best of our knowledge, byond lists are red-black trees combined with something else for associativity.

RemieRichards replied:
@Miauw62 R/Bs+Dynamic(expanding) Arrays, R/Bs are only used for associations though, p.sure.

Giacom commented on code/__DATASTRUCTURES/priority_queue.dm:
```diff
@@ -0,0 +1,65 @@
+
+//an ordered list, using the cmp proc to weight the list elements
+/PriorityQueue
+	var/list/L //the actual queue
```
Ideally a queue would use a linked list for the benefits of being able to remove/add the front element with O(1) efficiency.

Not sure how that would work in BYOND, you will need to do benchmarks. I am also assuming BYOND lists are dynamic arrays, they could be linked lists for all I know.

Giacom commented on code/__DATASTRUCTURES/priority_queue.dm:
```diff
@@ -0,0 +1,65 @@
+//////////////////////
+
+//an ordered list, using the cmp proc to weight the list elements
+/PriorityQueue
```
Ideally a queue would use a linked list for the benefits of being able to remove/add the front element with O(1) efficiency.

Not sure how that would work in BYOND, you will need to do benchmarks. I am also assuming BYOND lists are dynamic arrays, they could be linked lists for all I know.

RemieRichards replied:
I'll look into that for a future PR, I'm not fully familiar with Priority Queues personally, just moving their contents to the (now) best place, the datastructure folder.

Giacom commented on code/__DATASTRUCTURES/stacks.dm:
```diff
@@ -0,0 +1,55 @@
+	if(is_empty())
+		return STACK_UNDERFLOW
+	. = stack[stack.len]
+	stack.Cut(stack.len,0)
```
Not sure if it is better but you can do this instead:

```C++
stack.len -= 1
```

RemieRichards replied:
P.sure there have been troubles in the past of ```len--`` or equivalent syntax failing to actually destroy the element.
```Cut()``` itself is faster than ```Remove()```, so it's decently fast already.

Giacom commented on code/__DATASTRUCTURES/stacks.dm:
```diff
@@ -0,0 +1,55 @@
+	. = stack[stack.len]
+
+/datum/stack/proc/is_empty()
+	. = stack.len ? 0 : 1
```
Probably more clear to do

```C++
return stack.len > 0
```

RemieRichards replied:
I went for speed here, no idea if ```>``` evaluates faster than ```?``` but ```.``` is certainly slightly faster than ```return```

RemieRichards commented:
Reimplemented Linked Lists per @Giacom and @MrStonedOne 's comments
CPU for Linked Lists has dropped from 20/10 to 4/4 on Add/Remove which is an added bonus.

Giacom commented on code/__DATASTRUCTURES/linked_lists.dm:
```diff
@@ -0,0 +1,185 @@
+//Ok so it's technically a double linked list, bite me.
+
+/datum/linked_list
+	var/datum/linked_element/head
```
I think the generally agreed upon term is "node".

Giacom commented on code/__DATASTRUCTURES/linked_lists.dm:
```diff
@@ -0,0 +1,185 @@
+
+
+/datum/linked_list/proc/is_empty()
+	. = element_amt ? 0 : 1
```
```C++
. = element_amt <= 0
```

Makes more sense.

Giacom commented on code/__DATASTRUCTURES/stacks.dm:
```diff
@@ -0,0 +1,55 @@
+//Both count as failures, and they don't equate to each other
+//this lets us do if(pop()) without having to specifically check for underflow
+//same for if(push()) and overflow
+#define STACK_OVERFLOW	""
```
I think it would be clearer if you made the DEFINEs into negative numbers. They would still count as false and it wouldn't be as confusing.

RemieRichards replied:
That's true yes, at the time I forgot negative values equated to false

Giacom commented:
If you or anyone else wants to implement more algorithms, or look through them to see which could help with an upcoming project, take a look at this.

https://github.com/aalhour/C-Sharp-Algorithms

It is in C# and there are a lot of generics and interfaces but you can strip them out.

MrStonedOne commented:
Just a note, since its buried in a diff commit, this name conflicts with the stack datum used by ntsl.

Odds are, ntsl can be made to use this stack datum

RemieRichards commented:
@MrStonedOne If I rename my Push() and Pop() to be capitalised, and implement NTSL stack's Copy() they'd be 100% compatible (Mine has some extra functions so mine would absorb the NTSL one, not to mention NTSL is generally considered kill these days)

@Giacom thanks, I'll keep a note of it.

RemieRichards commented:
News just in, Relative pathing still a shit.
I had to correct all the ```/stack``` to ```/datum/stack``` (despite them being the same) due to Relative pathing not recognising that /stack is shorthand for /datum/stack,  *sigh*

* ```/datum/stack``` had absorbed and assumed the job of ```/stack```
* Rewrote the naming scheme of ```linked_list``` to use ```linked_node```s instead of ```linked_element```s
* Changed the case for all the data structures to match, primarily because it was required for NTSL and also because I couldn't stand them not being uniform

RemieRichards committed:
/datum/stack replaces /stack, naming scheme redo for procs and linked_element --> linked_node

Giacom commented:
You should look into whether BYOND utilizes caching with its bytecode. Linked lists are not really favoured due to the fact the CPU can fetch batches of data from RAM; this can cause issue with linked lists since their location in RAM is not guaranteed to be next to each other, while arrays and other data structures which have an underlying array will guarantee it, at the cost of having to allocate another array when the existing one gets too small.

MrStonedOne commented:
>You should look into whether BYOND utilizes caching with its bytecode. Linked lists are not really favoured due to the fact the CPU can fetch batches of data from RAM;

Tobba once said to me that not only did it not, the way it accessed memory made for extremely poor caching performance for what little caching could still be done by the cpu.

Chnkr commented:
> You should look into whether BYOND utilizes caching with its bytecode

It does not, and as @MrStonedOne explained earlier the way it handles lists() internally makes for extremely poor performance during runtime, i suggest you implement this somewhere in-game, and profile it that way.

Giacom commented:
Well that sucks.

Razharas commented:
Will you use it for something in the end rr?
Also please dont let the usual fate get to this pr

RemieRichards commented:
@Razharas WHAT USUAL FATE? Besides things in byond I can't control, it's fine to merge, has been for days.
